### DISCUSSION TITLE
Conditional generation of containers

### DISCUSSION THREAD

**Discussion by blandaadrian** [QuestPDF/QuestPDF#143] [Q&A] [answered]
Hello – we love QuestPDF, awesome Tool, very useful for our Project. 
I ran into a problem, maybe it's just my misunderstanding of the framework or Fluent Interface / C# in general, but here it is:
_______
I want to conditionally create containers (table cells) based on certain values like so:

``` C#
switch (style)
{
    case ReportStyle.Simple:
        {
            if (!isFirst)
            {
                container.BorderLeft(1);
            }
            container.BorderBottom(1);
        }
        break;

    case ReportStyle.Excel:
        {
            container.Border(1).Background(Colors.Indigo.Lighten5);
        }
        break;
} 
```

Now, the essential problem seems to be that this works:

``` C#
container.BorderLeft(1).BorderBottom(1).Text("Hello");
```
while this doesn't:
``` C#
container.BorderLeft(1);
container.BorderBottom(1);
container.Text("Hello");
```

It throws the following `DocumentComposeException: You should not assign multiple child elements to a single-child container.`

_______

I understand that this kind of usage may not be intended or may cause performance issues.
And I know of the alternative to create a component for every style, but I have more complex conditional elements than the example above. For those, a component for every single one of the possible combinations is not practical.

Am I doing it wrong? Is there another way to achieve this? 

Thanks alot!

**Comment by bennetbo** [comment, marked as answer]
Hey, in order to understand why that exception gets thrown, let me go through what structure QuestPDF actually generates for your code.

```csharp
container.BorderLeft(1).BorderBottom(1).Text("Hello");
```

This does work because each call, for example "BorderLeft()", returns a new container which can have another child.
The structure that is generated looks like this (Indent means its the child of the element above):

```yaml
- container
    - Border (left: 1)
        - Border(bottom: 1)
            - Text("Hello")
```

Actually due to optimization the structure looks more like this:

```yaml
- container
    - Border (left: 1, bottom: 1)
        - Text("Hello")
```
---
Now lets look at your example that doesnt work:

Step 1
```csharp
container.BorderLeft(1)
```

```yaml
- container
    - Border (left: 1)
```
Step 2

```csharp
container.BorderBottom(1)
```
```yaml
- container # Same container as above
    - Border(left: 1) # Already has child, conflict!
```
Which causes the `DocumentComposeException`.

---
Here is an example that should work.

```csharp
container
    .Element(c =>
    {
        switch (style)
        {
          case ReportStyle.Simple:
            if (!isFirst)
              return c.BorderLeft(1).BorderBottom(1);
            return c.BorderBottom(1);
          case ReportStyle.Excel:
            return c.Border(1).Background(Colors.Indigo.Lighten5);
          default: throw new NotImplementedException(nameof(style));
        }
    })
    .Text("Hello");

//Or

switch(style)
{
    case ReportStyle.Simple:
        if(!isFirst)
            container = container.BorderLeft(1);
        container = containter.BorderBottom(1);
        break;

    case ReportStyle.Excel:
        container = container.Border(1).Background(Colors.Indigo.Lighten5);
        break;
}

container.Text("Hello");
```

**Reply by blandaadrian** [reply]
I see. Thanks for the extremely quick and detailed answer, this helps a lot!
Kindest regards from Austria.

**Reply by MarcinZiabek** [reply]
@Bebo-Maker I am impressed by your precise explanation! Thank you 😁

**Comment by DawidRobertNowak** [comment]
Hello,

We are closing this thread due to inactivity. Thank you for taking the time to reach out and contribute to the project — your feedback helps us improve QuestPDF.

Should you encounter any further problems, please don't hesitate to open a new issue.

Thank you!
